src/index.ts

```diff
@@ -290,52 +318,56 @@ class WalletConnectProvider implements SignerProvider {
+    const parsedAccounts = accounts.map(parseAccount)
     if (this.sameAccounts(parsedAccounts, this.lastSetAccounts)) {
       return;
     } else {
       this.lastSetAccounts = parsedAccounts;
     }
 
-    const newAccounts = parsedAccounts.filter(account =>
-      isCompatibleChainGroup(account.group, this.chainGroup),
-    );
-    if (!this.sameAccounts(newAccounts, this.accounts)) {
+    const newAccounts = parsedAccounts
+      .filter(account =>
+        this.chains.includes(formatChain(this.networkId, account.group))
+      )
+      .filter((value, index, array) =>
+        array.findIndex(v => (v.address === value.address)) === index
+      )
+
+    if (newAccounts.length !== 0 && !this.sameAccounts(newAccounts, this.accounts)) {
       this.accounts = newAccounts;
-      this.events.emit(providerEvents.changed.accounts, newAccounts);
+      this.events.emit(PROVIDER_EVENTS.accountsChanged, newAccounts);
     }
   }
 }
 
 export function isCompatibleChain(chain: string): boolean {
-  return chain.startsWith(`${WalletConnectProvider.namespace}:`);
+  return chain.startsWith(`${ALEPHIUM_NAMESPACE}:`);
 }
 
-export function formatChain(networkId: number, chainGroup?: number): string {
-  const chainGroupEncoded = chainGroup !== undefined ? chainGroup : -1;
-  return `${WalletConnectProvider.namespace}:${networkId}/${chainGroupEncoded}`;
+export function formatChain(networkId: number, chainGroup: number): string {
+  return `${ALEPHIUM_NAMESPACE}:${networkId}/${chainGroup}`;
 }
 
-export function isCompatibleChainGroup(chainGroup: number, expectedChainGroup?: number): boolean {
-  return expectedChainGroup === undefined || expectedChainGroup === chainGroup;
+export function isCompatibleChainGroup(chainGroup: number, expectedChainGroup: number): boolean {
+  return expectedChainGroup === -1 || expectedChainGroup === chainGroup;
 }
 
-export function parseChain(chainString: string): [number, number | undefined] {
-  const [namespace, networkId, chainGroup] = chainString.replace(/\//g, ":").split(":");
+export function parseChain(chainString: string): [number, ChainGroup] {
+  const [_namespace, networkId, chainGroup] = chainString.replace(/\//g, ":").split(":");
   const chainGroupDecoded = parseInt(chainGroup, 10);
-  return [parseInt(networkId, 10), chainGroupDecoded === -1 ? undefined : chainGroupDecoded];
+  return [parseInt(networkId, 10), chainGroupDecoded as ChainGroup];
 }
 
 export function formatAccount(permittedChain: string, account: Account): string {
   return `${permittedChain}:${account.address}+${account.publicKey}`;
 }
 
 export function parseAccount(account: string): Account {
-  const [namespace, permittedChain, address, publicKey] = account.replace(/\+/g, ":").split(":");
+  const [_namespace, _networkId, group, address, publicKey] = account.replace(/\+|\//g, ":").split(":");
   return {
     address: address,
     publicKey: publicKey,
-    group: groupOfAddress(address),
+    group: +group
```

Comment:
My initial thought was that we should trust as less data as possible, that's why the group is calculated from address.

But I should have only relayed public key and calculate both address and public key from it. WDYT?

Reply:
Yeah, that is a good point, lets do that :+1: